pyteal/ast/abi/address_test.py

```diff
@@ -68,15 +68,15 @@ def test_Address_decode():
                         )
                     continue
 
-                expr = value.decode(
-                    encoded, startIndex=startIndex, endIndex=endIndex, length=length
-                )
+                expr = value.decode(encoded, startIndex=startIndex)
                 assert expr.type_of() == pt.TealType.none
                 assert expr.has_return() is False
 
                 expectedExpr = value.stored_value.store(
                     substringForDecoding(
-                        encoded, startIndex=startIndex, endIndex=endIndex, length=length
+                        encoded,
+                        startIndex=startIndex,
+                        length=value.length(),
```

Comment:
I'm getting different compilation results for this branch compared to mine. For example, when compiling [this PyTEAL](https://github.com/algorand/pyteal/blob/1ee615117da7617078fa2d68ad8546966a734eb7/tests/unit/blackbox_test.py#L124):

```python
@Blackbox(input_types=[None, pt.TealType.uint64, None])
@pt.ABIReturnSubroutine
def fn_3mixed_args_0ret(
    a: pt.abi.Uint64, b: pt.ScratchVar, C: pt.abi.StaticArray[pt.abi.Byte, Literal[10]]
) -> pt.Expr:
    return pt.Return()
```

I get the following change in compilation output:
```
#pragma version 6
txna ApplicationArgs 0
btoi
store 3
txna ApplicationArgs 1
btoi
store 4
txna ApplicationArgs 2
extract 0 10          <----------- NEW ADDITIONAL LINE (everything else the same)
store 5
load 3
int 4
load 5
callsub fn3mixedargs0ret_0
int 1
return

// fn_3mixed_args_0ret
fn3mixedargs0ret_0:
store 2
store 1
store 0
retsub
```

Reply:
Not sure why yet, or the impact, but I wanted to bring it to your attention.

Reply:
I mentioned this in the summary above. It may make sense to put this on the dev to specify a length if necessary but also seems incorrect to allow `decode` to read in more bytes than should be allowed given the types.

Reply:
Ok, I see. I'll have to think about the implications some more. Meanwhile, some of the roundtrip graviton tests on my branch are breaking. These might be fixed if I specify the extract length, but my first hunch is that I was just doing vanilla operations and am surprised by the break. It's hard to show the "pyteal" because it's auto generated, but here's the teal (approximately) that it generated, and that teal gives an error that indicates that the expected log wasn't generated.
```
00:'#pragma version 6'
01:'txna ApplicationArgs 0'
02:'extract 0 1'
03:'store 10'
04:'load 10'
05:'callsub roundtripper_2'
06:'store 0'
07:'byte 0x151F7C75'
08:'load 0'
09:'concat'
10:'log'
11:'int 1'
12:'return'
13:''
14:'// numerical_comp'
15:'numericalcomp_0:'
16:'store 8'
17:'int 18446744073709551615'
18:'load 8'
19:'-'
20:'store 9'
21:'load 9'
22:'retsub'
```

Reply:
Ah, length is just number of elements not byte length I guess?

Reply:
yep

Reply:
When I fixed locally on my branch all the graviton tests pass, except the most ambitious. I get the following error:

```
>           raise TealInputError(
                "Expected None for endIndex and length on static array"
            )
E           pyteal.TealInputError: Expected None for endIndex and length on static array

pyteal/ast/abi/array_static.py:88: TealInputError
```
Not sure if I'm running out of gas in the dry run, or there is some deeper issue.

Reply:
I think there are two things in here, `length` is more likely to be static array element number, while `byte_length` might be the static length of the byte encoding (can precompute for static type).

Reply:
I create a temp branch showing the [failing test](https://github.com/algorand/pyteal/runs/6479500080?check_suite_focus=true#step:8:100).

Reply:
@tzaffi can you give it another shot with the latest changes?

Reply:
It's still failing, but this time I think I know why. Consider the following ABI type which is failing:

```python
        abi.DynamicArray[
            abi.Tuple4[
                abi.StaticArray[abi.Byte, Literal[4]],
                abi.Tuple2[abi.Bool, abi.Bool],
                abi.Uint64,
                abi.Address,
            ]
        ],
```
In particular, we have a `Tuple4[StaticArray[...]]`.  If we have instance `w, x: ...` of this type and we set one to the other `w.set(x)` eventually we hit [this line in tuple::indexTuple()](https://github.com/algorand/pyteal/blob/d21678e7444cd4e4481c7ccd0e7f3c641fb269f7/pyteal/ast/abi/tuple.py#L201) which supplies a non-empty `length` to `x.decode()` and results in the error above [at this line](https://github.com/algorand/pyteal/blob/4c1d4cd46d1eb5d258b9bd11db2218f42a5277c4/pyteal/ast/abi/array_static.py#L94). 

So I suspect that the check for `endIndex == None == length` isn't quite right.

Reply:
oh... then there is indeed something magical behind these small code changes...

Reply:
I have a minor conjecture (or suggestion):

what if we don't enforce that `decode` in `StaticArray` to raise error in `endIndex is None or length is None`, then we just do the decoding like its parent `ArrayBase` and I guess the behavior is preserved?

Reply:
I can remove the error thrown on endIndex and pass it into decode but I'm torn on the length check since we set it directly it'd seem strange to allow length argument but not use it?